### PR TITLE
Implemented step 1 and 2 for WoT. Still need to notify the POR subjec…

### PR DESCRIPTION
**Pull request by ykocaogullar**
Implemented following functionalities:

- Fetch stranger certificate by keylocator
- Request POR certificate issued by every other known (authenticated) user
  - If my cert store has N other users, send N interests for each
  - If not getting at least one POR, abort (treat as data verification failure)
- Cross-sign stranger public key as POR certificate
- Putting POR certificate into cert store
- POR subject serves all POR it received

Missing functionality: Notify the POR subject a new POR is generated, let it fetch by name

### CODE REVIEW

**Comment by zjkmxy** [comment]
Hello. Please run `pnpm format`, `pnpm lint` and `pnpm test` locally and fix the errors.

**Comment by zjkmxy** [comment]
How is your progress on this?

**Comment by ykocaogullar** [comment]
> How is your progress on this?

I'll fix these before the end of the weekend.

File: src/security/cert-storage.ts [hunk 1]
```diff
@@ -135,17 +136,92 @@ export class CertStorage implements SecurityAgent {
...
+    for (const item of this.trustedNames) {
+      // Construct <Z_a name encoded>
+      const parts: string[] = item.split('/');
+      let partsBeforeKey: string = '';
+      for (let i = 0; i < parts.length; i++) { //put <Z_a name encoded> together with a loop
+        if (parts[i].includes('KEY')) {
+          break;
+        }
+        partsBeforeKey += parts[i] + '/';
+      }
```
Comment (zjkmxy) [inline src/security/cert-storage.ts:167]: First, I don't think this is the correct behavior. I think `<Tb keyid>` is a **single** component containing the encoded name. Please confirm with @tianyuan129 .
If this behavior is correct, you can still get rid of the loop. See [Name.append](https://ndnts-docs.ndn.today/typedoc/classes/_ndn_packet.name-1#append)

File: src/security/cert-storage.ts [hunk 2]
```diff
@@ -135,17 +136,92 @@ export class CertStorage implements SecurityAgent {
...
+
+      // return if we found that one of the users we trust, also trusts the new user
+      const response: Certificate | undefined = await this.getCertificate(
+        new Name(strangerKeyname + partsBeforeKey + 'v1'),
+        false,
```
Comment (zjkmxy) [inline src/security/cert-storage.ts:172]: This means PoR can be recursively signed by other unknown certifcates. Please check if the behavior if desired.

File: src/security/cert-storage.ts [hunk 3]
```diff
@@ -135,17 +136,92 @@ export class CertStorage implements SecurityAgent {
...
+    strangerKeyname: string, // name of t_b, i.e. "/<𝑍𝑏 name>/KEY/<𝑇𝑏 keyid>"
+    certName: Name, // follow PoR naming format
+  ): Promise<Certificate> {
+    // Request POR certificate issued by every other known (authenticated) user
+    const foreignPoR: Certificate | undefined = await this.fetchPoR(strangerKeyname);
+
+    // If not getting at least one POR, abort (treat as data verification failure)
+    if (foreignPoR == undefined) {
+      throw new Error('Data verification failure: No POR certificate found issued by trusted user');
+    }
```
Comment (zjkmxy) [inline src/security/cert-storage.ts:193]: This reads as "a PoR cannot be generated without an existing one".
Then, how can the first PoR be generated?

File: src/security/cert-storage.ts [hunk 4]
```diff
@@ -135,17 +136,92 @@ export class CertStorage implements SecurityAgent {
...
+
+    // Get public key of C_b
+    const certificate: Certificate | undefined = await this.getCertificate(new Name(strangerKeyname), false);
```
Comment (zjkmxy) [inline src/security/cert-storage.ts:196]: Do you want to fetch a certificate when generating PoR?

File: src/security/cert-storage.ts [hunk 5]
```diff
@@ -135,17 +136,92 @@ export class CertStorage implements SecurityAgent {
...
+    if (!this._signer) {
+      throw new Error('Signer is undefined');
+    }
+    // Create a self-signed certificate, i.e. the PoR
+    const proofOfZoneRecognition = await Certificate.build({
+      name: certName,
+      validity: validityPeriod,
+      publicKeySpki: publicKey,
+      signer: this._signer,
+    });
```
Comment (zjkmxy) [inline src/security/cert-storage.ts:220]: Why a PoR is self-signed?

**Review by zjkmxy — CHANGES_REQUESTED** [review]

**Pull request closed without merging**